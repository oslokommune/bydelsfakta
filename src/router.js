import Vue from 'vue';
import Router from 'vue-router';

import { topicNames, disabledTopics } from './config/topics';
import allDistricts from './config/allDistricts';

const District = () => import('./views/District');
const Topic = () => import('./views/Topic');
const NotFound = () => import('./views/NotFound');

Vue.use(Router);

export const routes = [
  {
    path: '/',
    name: 'Home',
    component: District,
  },
  {
    path: '/bydel/:district?',
    name: 'District',
    component: District,
    props: true,
  },
  {
    path: '/bydel/:district/:topic',
    name: 'Topic',
    component: Topic,
    props: true,
  },
  {
    path: '*',
    name: 'NotFound',
    component: NotFound,
  },
];

const router = new Router({
  mode: 'history',
  routes,
});

// Temp while we find out what our homepage should look like
router.beforeEach((to, from, next) => {
  if (to.params.topic) {
    if (!topicNames.find(name => name === to.params.topic) || disabledTopics.includes(to.params.topic)) {
      next({ name: 'NotFound', params: [to.path] });
    }
  }

  if (to.params.district) {
    const districts = to.params.district.split('-');
    if (districts[0] === 'alle') {
      next();
    } else if (districts.length === 1) {
      if (allDistricts.find(district => district.uri === districts[0])) {
        next();
      } else if (allDistricts.find(district => district.key === districts[0])) {
        next();
      } else {
        next({ name: 'NotFound', params: [to.path] });
      }
    } else if (districts.length > 1) {
      const errors = [];
      districts.forEach(district => {
        if (!allDistricts.find(item => item.key === district)) {
          errors.push(district);
        }
      });
      if (errors.length > 0) {
        next({ name: 'NotFound', params: [to.path] });
      }
    }
  }

  if (to.params.topic !== from.params.topic || (to.name === 'Topic' && from.name === 'District')) {
    window.scroll(0, 0);
  }

  next();
});

export default router;
