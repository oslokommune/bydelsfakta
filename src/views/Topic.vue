<template>
  <div class="main-container">
    <div class="main-container__cards">
      <template v-for="(card, cardIndex) in topics[topic].cards">
        <graph-card
          :key="`card-${cardIndex}`"
          :settings="card"
          v-if="
            card.production === productionMode && card.production !== null
              ? true
              : card.production === true && productionMode === false
              ? true
              : productionMode === null
          "
        />
      </template>
      <div class="related">
        <h2 class="section-heading">{{ $t('topic.seeMore') }}</h2>
        <div class="topics-grid">
          <v-category
            v-for="(item, relatedIndex) in topics[`${topic}`].related"
            :key="`related-${relatedIndex}`"
            :id="topics[`${item}`].value"
            :category="topics[`${item}`].options.kategori"
            :topic="topics[`${item}`].options.tema"
            :bg-image="topics[`${item}`].options.bgImage"
            :txt-color="topics[`${item}`].options.txtColor"
            :district="district"
            :disabled="
              topics[`${item}`].production === productionMode && topics[`${item}`].production !== null
                ? false
                : topics[`${item}`].production === true && productionMode === false
                ? false
                : productionMode !== null
            "
          />
        </div>
      </div>
    </div>
  </div>
</template>

<script>
import { mapGetters, mapState } from 'vuex';
import VCategory from '../components/VCategory';
import GraphCard from '../components/GraphCard';

import { topics, disabledTopics } from '../config/topics';
import { getDistrictName, getHumanReadableTopic } from '../util';

export default {
  name: 'Topic',
  components: {
    GraphCard,
    VCategory,
  },
  props: {
    district: {
      type: String,
      required: true,
    },
    topic: {
      type: String,
      required: true,
    },
  },

  metaInfo() {
    return {
      title: `${getHumanReadableTopic(this.topic)} | ${
        this.compareDistricts ? 'Sammenlign bydeler' : getDistrictName(this.district)
      } | Bydelsfakta`,
    };
  },

  data() {
    return {
      topics: topics,
      disabledTopics: disabledTopics,
    };
  },

  computed: {
    ...mapState(['productionMode', 'compareDistricts']),
    ...mapGetters(['geoDistricts']),
  },
};
</script>

<style scoped>
.related {
  width: 100%;
}
</style>
