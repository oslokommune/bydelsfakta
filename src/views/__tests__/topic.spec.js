import 'whatwg-fetch';
import { createLocalVue, mount } from '@vue/test-utils';
import Vuex from 'vuex';
import VueResize from 'vue-resize';
import VueRouter from 'vue-router';

import Topic from '../Topic';
import clickOutside from '../../directives/clickOutside';
import store from '../../store';
import setupI18n from '../../i18n';
import { routes } from '../../router';

const i18n = setupI18n();

describe('Topic', () => {
  let wrapper = null;
  let router = null;

  beforeEach(() => {
    const localVue = createLocalVue();
    localVue.use(VueRouter);
    localVue.use(VueResize);
    localVue.use(Vuex);
    router = new VueRouter({ routes });
    localVue.directive('click-outside', clickOutside);
    wrapper = mount(Topic, {
      propsData: {
        district: 'bydel',
        topic: 'alder',
      },
      localVue,
      router,
      store,
      i18n,
      stubs: {
        'ok-icon': true,
        spinner: true,
        'v-category': true,
      },
    });
    router.push('/bydel/alle');
  });

  afterEach(() => {
    wrapper.destroy();
  });

  test('renders topic-component and finds main-container class', () => {
    expect(wrapper.classes('main-container')).toBe(true);
  });

  test('renders component correctly', () => {
    expect(wrapper.element).toMatchSnapshot();
  });
});
