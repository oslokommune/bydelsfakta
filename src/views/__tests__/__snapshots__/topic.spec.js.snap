// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Topic renders component correctly 1`] = `
<div
  class="main-container"
>
  <div
    class="main-container__cards"
  >
    <section
      class="card-container large"
    >
      <div
        class="card"
      >
        <!---->
         
        <header
          class="card__header"
        >
          <div
            class="card__headertext"
          >
            <h2
              class="card__title"
            >
              Befolkningen etter alder
            </h2>
          </div>
           
          <nav
            class="card__nav"
          >
            <div
              class="tabs"
              role="tablist"
            >
              <div
                class="resize-observer"
                data-v-b329ee4c=""
                tabindex="-1"
              >
                <object
                  aria-hidden="true"
                  data="about:blank"
                  tabindex="-1"
                  type="text/html"
                />
              </div>
               
              <button
                aria-label="Antall"
                class="tabs__tab active tabs__tab--hidden"
                data-id="tabButton-Befolkningen etter alder-0"
                role="tab"
              >
                Antall
              </button>
              <button
                aria-label="Andel"
                class="tabs__tab tabs__tab--hidden"
                data-id="tabButton-Befolkningen etter alder-1"
                role="tab"
              >
                Andel
              </button>
               
              <select
                aria-hidden=""
                class="select tabs__select"
              >
                <option
                  value="0"
                >
                  Antall
                </option>
                <option
                  value="1"
                >
                  Andel
                </option>
              </select>
            </div>
             
            <div
              class="card__toggle-menu"
            >
              <button
                aria-label="Graf"
                class="card__toggle-button card__toggle-button--active"
                title="Graf"
              >
                <ok-icon-stub
                  class="context-menu__dropdown-item-icon"
                  icon-ref="graph"
                  options="[object Object]"
                  stroke-width="2.5"
                />
              </button>
               
              <button
                aria-label="Tabell"
                class="card__toggle-button"
                title="Tabell"
              >
                <ok-icon-stub
                  class="context-menu__dropdown-item-icon"
                  icon-ref="data-table"
                  options="[object Object]"
                  stroke-width="2.5"
                />
              </button>
               
              <!---->
            </div>
             
            <div
              class="context-menu"
            >
              <button
                aria-haspopup="true"
                aria-label="Åpne meny"
                class="context-menu__button"
                data-id="context-menu-button"
                title="Åpne meny"
              >
                <ok-icon-stub
                  icon-ref="hamburger"
                />
                 
                <span
                  class="button-label"
                >
                  Valg
                </span>
              </button>
               
              <!---->
            </div>
          </nav>
        </header>
         
        <div
          class="graph__shadow"
          role="tabpanel"
        >
          <div
            aria-hidden="true"
            class="graph__container"
          >
            <div
              class="spinner"
            >
              <spinner-stub />
               
              <span
                class="spinner__text"
              >
                Henter data
              </span>
            </div>
             
            <!---->
             
            <svg
              aria-hidden="true"
              class="graph__svg loading"
            />
          </div>
           
          <div
            class="graph__tablecontainer visually-hidden"
          >
            <h3
              class="table-heading"
            />
             
            <table
              class=""
            >
              <thead />
               
              <tbody />
            </table>
             
            <!---->
          </div>
           
          <div
            class="resize-observer"
            data-v-b329ee4c=""
            tabindex="-1"
          >
            <object
              aria-hidden="true"
              data="about:blank"
              tabindex="-1"
              type="text/html"
            />
          </div>
           
          <div
            class="help"
          >
            <button
              aria-label="Vis hjelpetekst"
              class="help__button"
            >
              <ok-icon-stub
                icon-ref="help"
                options="[object Object]"
              />
            </button>
             
            <!---->
          </div>
        </div>
         
        <!---->
         
        <!---->
      </div>
    </section>
    <section
      class="card-container large"
    >
      <div
        class="card"
      >
        <!---->
         
        <header
          class="card__header"
        >
          <div
            class="card__headertext"
          >
            <h2
              class="card__title"
            >
              Befolkningspyramide
            </h2>
          </div>
           
          <nav
            class="card__nav"
          >
            <div
              class="tabs"
              role="tablist"
            >
              <div
                class="resize-observer"
                data-v-b329ee4c=""
                tabindex="-1"
              >
                <object
                  aria-hidden="true"
                  data="about:blank"
                  tabindex="-1"
                  type="text/html"
                />
              </div>
               
              <button
                aria-label="Status"
                class="tabs__tab active tabs__tab--hidden"
                data-id="tabButton-Befolkningspyramide-0"
                role="tab"
              >
                Status
              </button>
               
              <select
                aria-hidden=""
                class="select tabs__select"
              >
                <option
                  value="0"
                >
                  Status
                </option>
              </select>
            </div>
             
            <div
              class="card__toggle-menu"
            >
              <button
                aria-label="Graf"
                class="card__toggle-button card__toggle-button--active"
                title="Graf"
              >
                <ok-icon-stub
                  class="context-menu__dropdown-item-icon"
                  icon-ref="graph"
                  options="[object Object]"
                  stroke-width="2.5"
                />
              </button>
               
              <button
                aria-label="Tabell"
                class="card__toggle-button"
                title="Tabell"
              >
                <ok-icon-stub
                  class="context-menu__dropdown-item-icon"
                  icon-ref="data-table"
                  options="[object Object]"
                  stroke-width="2.5"
                />
              </button>
               
              <!---->
            </div>
             
            <div
              class="context-menu"
            >
              <button
                aria-haspopup="true"
                aria-label="Åpne meny"
                class="context-menu__button"
                data-id="context-menu-button"
                title="Åpne meny"
              >
                <ok-icon-stub
                  icon-ref="hamburger"
                />
                 
                <span
                  class="button-label"
                >
                  Valg
                </span>
              </button>
               
              <!---->
            </div>
          </nav>
        </header>
         
        <div
          class="graph__shadow"
          role="tabpanel"
        >
          <div
            aria-hidden="true"
            class="graph__container"
          >
            <div
              class="spinner"
            >
              <spinner-stub />
               
              <span
                class="spinner__text"
              >
                Henter data
              </span>
            </div>
             
            <!---->
             
            <svg
              aria-hidden="true"
              class="graph__svg loading"
            />
          </div>
           
          <div
            class="graph__tablecontainer visually-hidden"
          >
            <h3
              class="table-heading"
            />
             
            <table
              class=""
            >
              <thead />
               
              <tbody />
            </table>
             
            <!---->
          </div>
           
          <div
            class="resize-observer"
            data-v-b329ee4c=""
            tabindex="-1"
          >
            <object
              aria-hidden="true"
              data="about:blank"
              tabindex="-1"
              type="text/html"
            />
          </div>
           
          <!---->
        </div>
         
        <!---->
         
        <!---->
      </div>
    </section>
    <section
      class="card-container large"
    >
      <div
        class="card"
      >
        <!---->
         
        <header
          class="card__header"
        >
          <div
            class="card__headertext"
          >
            <h2
              class="card__title"
            >
              Gjennomsnitt- og medianalder
            </h2>
          </div>
           
          <nav
            class="card__nav"
          >
            <div
              class="tabs"
              role="tablist"
            >
              <div
                class="resize-observer"
                data-v-b329ee4c=""
                tabindex="-1"
              >
                <object
                  aria-hidden="true"
                  data="about:blank"
                  tabindex="-1"
                  type="text/html"
                />
              </div>
               
              <button
                aria-label="Status"
                class="tabs__tab active tabs__tab--hidden"
                data-id="tabButton-Gjennomsnitt- og medianalder-0"
                role="tab"
              >
                Status
              </button>
               
              <select
                aria-hidden=""
                class="select tabs__select"
              >
                <option
                  value="0"
                >
                  Status
                </option>
              </select>
            </div>
             
            <div
              class="card__toggle-menu"
            >
              <button
                aria-label="Graf"
                class="card__toggle-button card__toggle-button--active"
                title="Graf"
              >
                <ok-icon-stub
                  class="context-menu__dropdown-item-icon"
                  icon-ref="graph"
                  options="[object Object]"
                  stroke-width="2.5"
                />
              </button>
               
              <button
                aria-label="Tabell"
                class="card__toggle-button"
                title="Tabell"
              >
                <ok-icon-stub
                  class="context-menu__dropdown-item-icon"
                  icon-ref="data-table"
                  options="[object Object]"
                  stroke-width="2.5"
                />
              </button>
               
              <button
                aria-label="Kart"
                class="card__toggle-button"
                title="Kart"
              >
                <ok-icon-stub
                  class="context-menu__dropdown-item-icon"
                  icon-ref="map-pin"
                  options="[object Object]"
                  stroke-width="2.5"
                />
              </button>
            </div>
             
            <div
              class="context-menu"
            >
              <button
                aria-haspopup="true"
                aria-label="Åpne meny"
                class="context-menu__button"
                data-id="context-menu-button"
                title="Åpne meny"
              >
                <ok-icon-stub
                  icon-ref="hamburger"
                />
                 
                <span
                  class="button-label"
                >
                  Valg
                </span>
              </button>
               
              <!---->
            </div>
          </nav>
        </header>
         
        <div
          class="graph__shadow"
          role="tabpanel"
        >
          <div
            aria-hidden="true"
            class="graph__container"
          >
            <div
              class="spinner"
            >
              <spinner-stub />
               
              <span
                class="spinner__text"
              >
                Henter data
              </span>
            </div>
             
            <!---->
             
            <svg
              aria-hidden="true"
              class="graph__svg loading"
            />
          </div>
           
          <div
            class="graph__tablecontainer visually-hidden"
          >
            <h3
              class="table-heading"
            />
             
            <table
              class=""
            >
              <thead />
               
              <tbody />
            </table>
             
            <!---->
          </div>
           
          <div
            class="resize-observer"
            data-v-b329ee4c=""
            tabindex="-1"
          >
            <object
              aria-hidden="true"
              data="about:blank"
              tabindex="-1"
              type="text/html"
            />
          </div>
           
          <!---->
        </div>
         
        <!---->
         
        <!---->
      </div>
    </section>
     
    <div
      class="related"
    >
      <h2
        class="section-heading"
      >
        Se også
      </h2>
       
      <div
        class="topics-grid"
      >
        <v-category-stub
          bgimage="/img/folkemengde"
          category="Befolkning"
          district="gamleoslo"
          id="befolkningsutvikling"
          topic="Befolkningsutvikling"
          txtcolor="rgb(245, 173, 165)"
        />
        <v-category-stub
          bgimage="/img/husholdninger"
          category="Befolkning"
          district="gamleoslo"
          id="husholdninger"
          topic="Husholdninger"
          txtcolor="rgb(245, 173, 165)"
        />
        <v-category-stub
          bgimage="/img/levekaar"
          category="Levekår"
          district="gamleoslo"
          id="levekaar"
          topic="Levekårs&shy;indikatorer"
          txtcolor="rgb(244, 206, 138)"
        />
      </div>
    </div>
  </div>
</div>
`;
